lib/player.js

```diff
@@ -3374,28 +3405,86 @@ shaka.Player = class extends shaka.util.FakeEventTarget {
     }
   }
 
-
   /**
    * Construct and fire a Player.Metadata event
    *
+   * @param {shaka.extern.MetadataTimelineRegionInfo} region
+   * @private
+   */
+  dispatchMetadataEvent_(region) {
+    const eventName = shaka.util.FakeEvent.EventName.Metadata;
+    const data = new Map()
+        .set('startTime', region.startTime)
+        .set('endTime', region.endTime)
+        .set('metadataType', region.schemeIdUri)
+        .set('payload', region.payload);
+    this.dispatchEvent(shaka.Player.makeEvent_(eventName, data));
+  }
+
+
+  /**
+   * Add metadata to region timeline
+   *
    * @param {number} startTime
    * @param {?number} endTime
    * @param {string} metadataType
    * @param {shaka.extern.MetadataFrame} payload
    * @private
    */
-  dispatchMetadataEvent_(startTime, endTime, metadataType, payload) {
+  addMetadataToRegionTimeline_(startTime, endTime, metadataType, payload) {
+    if (!this.metadataRegionTimeline_) {
+      return;
+    }
     goog.asserts.assert(!endTime || startTime <= endTime,
         'Metadata start time should be less or equal to the end time!');
-    const eventName = shaka.util.FakeEvent.EventName.Metadata;
-    const data = new Map()
-        .set('startTime', startTime)
-        .set('endTime', endTime)
-        .set('metadataType', metadataType)
-        .set('payload', payload);
+    /** @type {shaka.extern.MetadataTimelineRegionInfo} */
+    const region = {
+      schemeIdUri: metadataType,
+      startTime,
+      endTime: endTime || Infinity,
+      id: '',
+      payload,
+    };
+    // JSON stringify produces a good ID in this case.
+    region.id = JSON.stringify(region);
```

Comment:
this way, id will always contain `"id":""`, is this expected?

Reply:
Yes, it is expected because we have not yet generated the ID, the ID is generated here.